docker-java-core/src/main/java/com/github/dockerjava/core/FramedInputStreamConsumer.java

```diff
@@ -0,0 +1,99 @@
+package com.github.dockerjava.core;
+
+import com.github.dockerjava.api.async.ResultCallback;
+import com.github.dockerjava.api.model.Frame;
+import com.github.dockerjava.api.model.StreamType;
+
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.function.Consumer;
+
+class FramedInputStreamConsumer implements Consumer<DockerHttpClient.Response> {
+
+    private final ResultCallback<Frame> resultCallback;
+
+    FramedInputStreamConsumer(ResultCallback<Frame> resultCallback) {
+        this.resultCallback = resultCallback;
+    }
+
+    @Override
+    public void accept(DockerHttpClient.Response response) {
+        try {
+            InputStream body = response.getBody();
+
+            byte[] buffer = new byte[1024];
+            while (true) {
+                // See https://docs.docker.com/engine/api/v1.37/#operation/ContainerAttach
+                // [8]byte{STREAM_TYPE, 0, 0, 0, SIZE1, SIZE2, SIZE3, SIZE4}[]byte{OUTPUT}
+
+                int streamTypeByte = body.read();
+                if (streamTypeByte < 0) {
+                    return;
+                }
+
+                StreamType streamType = streamType(streamTypeByte);
+
+                if (streamType == StreamType.RAW) {
+                    resultCallback.onNext(new Frame(StreamType.RAW, new byte[] { (byte) streamTypeByte }));
+
+                    int readBytes;
+                    while ((readBytes = body.read(buffer)) >= 0) {
+                        if (readBytes == buffer.length) {
+                            resultCallback.onNext(new Frame(StreamType.RAW, buffer));
+                        } else {
+                            resultCallback.onNext(new Frame(StreamType.RAW, Arrays.copyOf(buffer, readBytes)));
+                        }
+                    }
```

Comment:
regarding my love for private methods I would make this private method `readRawFrame(body)` (or even make body an instance variable?)

Reply:
we would also need to pass `buffer` (to reuse it) and overall I would prefer to keep this algorithm in one method, to make it easier to follow